Aws organizations

p0tr3c opened the pull request:
##### SUMMARY
Adds ability to create, delete aws organizational units

##### ISSUE TYPE
- Feature Pull Request
- New Module Pull Request

##### COMPONENT NAME

aws_organizations

##### ADDITIONAL INFORMATION
Sample playbook use
```
- hosts: localhost
  connection: local
  gather_facts: false
  vars:
    org_ous:
      - Prod
      - Prod/Infra
      - Prod/Infra/Sec
      - Prod/Sec
      - Dev
      - Dev/Infra
      - Sec
  tasks:
    - name: Create AWS Organizational Unit
      aws_organizations:
        name: "{{ item }}"
        state: present
      loop: "{{ org_ous }}"
      tags:
        - create
    - name: Get information about organizational unit
      aws_organizations:
        name: Prod
        state: describe
      register: prod_ou_info
    - name: List all AWS Organizational Units
      aws_organizations:
        state: list
      register: organization_tree_info
      tags:
        - list
    - name: Delete AWS Organization Unit
      aws_organizations:
        name: "{{ item }}"
        state: absent
      loop: "{{ org_ous | reverse | list }}"
      tags:
        - delete
```

p0tr3c committed:
Basic organizations management module

p0tr3c committed:
Fix invalid version_added

p0tr3c committed:
Add missing quotes

flowerysong commented on lib/ansible/modules/cloud/amazon/aws_organizations.py:
```diff
@@ -0,0 +1,312 @@
+      - present
+      - absent
+      - describe
+      - list
```
While you may have encountered existing modules that combine functionality like this, it is not allowed in new modules (see https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html#scoping-your-module-s). list/describe should be split into a separate _info module.

p0tr3c replied:
Thanks for quick feedback. I will remove the info functionality, thank you for pointing it out.
I've seen a pull request which proposes _info module (#55837), so I won't create a new one at the time.

p0tr3c committed:
Remove info/describe option

p0tr3c committed:
Fix requried flag on 'name' argument

p0tr3c committed:
Remove incorrectly set check_mode

p0tr3c committed:
Add simple check_mode support

p0tr3c committed:
Remove wraper

tremble commented on lib/ansible/modules/cloud/amazon/aws_organizations.py:
```diff
@@ -0,0 +1,290 @@
+module: aws_organizations
+short_description: Manages AWS Organization
+description:
+  - Creates and deletes AWS OrganizationalUnits
```
Nit: please see https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#documentation-block

> Descriptions should always start with a capital letter and end with a full stop. Consistency always helps.

```suggestion
  - Creates and deletes AWS OrganizationalUnits.
```

p0tr3c replied:
Fixed

tremble commented on lib/ansible/modules/cloud/amazon/aws_organizations.py:
```diff
@@ -0,0 +1,290 @@
+
+
+DOCUMENTATION = """
+module: aws_organizations
```
Why organizations rather than aws_organization_units?

p0tr3c replied:
No good reason, so renamed

tremble commented on lib/ansible/modules/cloud/amazon/aws_organizations.py:
```diff
@@ -0,0 +1,290 @@
+
+try:
+    import boto3
+    import botocore.exceptions
```
```suggestion
```

Not sure we need this

p0tr3c replied:
Cannot see the need for it, removed

tremble commented on lib/ansible/modules/cloud/amazon/aws_organizations.py:
```diff
@@ -0,0 +1,290 @@
+                children_ous += page['OrganizationalUnits']
+        return children_ous
+
+    def get_children_accounts(self, parent_id):
```
This may be better named 'get_child_accounts'

p0tr3c replied:
Agree, renamed

tremble commented on lib/ansible/modules/cloud/amazon/aws_organizations.py:
```diff
@@ -0,0 +1,290 @@
+        result['state'] = 'present'
+        result['ou'] = ou
+
+    if module.check_mode:
```
This logic _feels_ wrong...  The check_mode code path should be as close as possible to non-check_mode code path, without actually making the change.

I think it would be better to move the check_mode logic inside the  "if ou is (not) None:" branches.

```
result['changed'] = True
if module.check_mode:
  client.create_ou(ou_name)
  result['state'] = 'present'
```

p0tr3c replied:
Fixed

tremble commented on lib/ansible/modules/cloud/amazon/aws_organizations.py:
```diff
@@ -0,0 +1,290 @@
+
+    def create_ou(self, name):
+        ou_name = name.rstrip('/').lstrip('/')
+        ou_path = ou_name.split('/')
```
This feels like hidden magic (and isn't documented).

It's very similar to the underlying ARN logic, but replaces IDs with "Friendly" names.

p0tr3c replied:
Required to handle full OU paths, and handle user input "/root/parent/child" or "root/parent/child" or (for unknown reason) "/root/parent/child/". Comments added to the code

tremble reviewed (changes requested):
1. I really don't like the logic around parenting.
    1. Parent should probably be an option on its own taking an ARN, but possibly allowing the "/" notation with Friendly names.
    2. The current logic's not documented.
2. Since AWS supports changing the Friendly Name, my personal preference is that 'ID' should be an additional option, and we should support changing the Friendly name.
3. The code path around check_mode looks too detached from the non-check_mode code path.  This will potentially result in problems in future.

p0tr3c committed:
Resolve comments from tremble
- Update description
- Rename module to limit it to organization units
- Remove unnecessary import
- Rename get_children_accounts function to get_child_accounts
- Refactor check module to be closer to the non-check code
- Comment the recursive OU name handling code

- Rename the module file

p0tr3c commented:
@tremble  thanks for review. I've addressed all the comments in the latest commit.

tremble commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -202,6 +201,10 @@ def _create_ou(self, name, parent_id):
+    # leading, and ending "/" will be stripped to handle
+    # "/root/child" or "root/parent/" inputs
+    # Parent ous will be recursively dereferenced
     def create_ou(self, name):
```
Rather than adding this as a comment here, it would be really good if you could add it to the description up on line 59.

It's obvious to you, but if I used the ARN, it wouldn't work...

tremble replied:
Would be nice if you accepted ARNs too ;)

p0tr3c replied:
Let me see what I can do

p0tr3c replied:
I've refactored the code and added support of ARN

tremble commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -255,27 +258,21 @@ def main():
+            result['changed'] = False
+        else:
+            if module.check_mode:
                 result['changed'] = True
```
Just to reduce duplication, you could switch the 'else' below to a ``if not module.check_mode:`` and leave the ``result['changed'] = True`` outside the if statement.  the result won't be returned if an exception is thrown while making the change.   (and the same below)

tremble replied:
Realised I've explained poorly:

```
if ou is None:
  result['changed'] = False
else:
  result['changed'] = True
  if not module.check_mode:
    client.delete_ou(ou['Id'])
    result['state'] = 'absent'
```

p0tr3c replied:
This looks much better, thanks

tremble commented:
While officially we 'require' unit tests.  I'm not sure it's practical in this case.

tremble commented:
I asked about testing on IRC and @s-hertel suggested the following:

> It could be unit tested  with API mocking (or integration tested, similar to https://github.com/ansible/ansible/tree/devel/test/integration/targets/inventory_aws_conformance). Less useful but still helpful for predictable output.

p0tr3c commented:
I've also come across https://github.com/spulec/moto library. But I don't believe it supports organizations yet. I will double check.

p0tr3c committed:
Refactor to remove duplicate code in check_mode

p0tr3c committed:
Document the full OU path handling in the description of name parameter

p0tr3c committed:
Move the description of the input name parameter to correct location

tremble commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,284 @@
+options:
+  name:
+    description:
+      - Name of the organizational unit
```
```suggestion
      - Name of the organizational unit.
```

tremble commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,284 @@
+    type: str
+  state:
+    description:
+      - State of the organizational unit
```
```suggestion
      - State of the organizational unit.
```

tremble commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,284 @@
+    pass
+
+from ansible.module_utils.ec2 import AWSRetry, boto3_tag_list_to_ansible_dict, ansible_dict_to_boto3_tag_list, camel_dict_to_snake_dict
+from ansible.module_utils.aws.core import AnsibleAWSModule, is_boto3_error_code
```
```suggestion
from ansible.module_utils.aws.core import AnsibleAWSModule
```

tremble commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,284 @@
+
+from ansible.module_utils.ec2 import AWSRetry, boto3_tag_list_to_ansible_dict, ansible_dict_to_boto3_tag_list, camel_dict_to_snake_dict
+from ansible.module_utils.aws.core import AnsibleAWSModule, is_boto3_error_code
+from ansible.module_utils._text import to_native
```
```suggestion
```

tremble commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,284 @@
+    # handled by AnsibleAWSModule
+    pass
+
+from ansible.module_utils.ec2 import AWSRetry, boto3_tag_list_to_ansible_dict, ansible_dict_to_boto3_tag_list, camel_dict_to_snake_dict
```
```suggestion
```

tremble commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,284 @@
+class AwsOrganizations():
+    def __init__(self, module):
+        self.module = module
+        try:
```
I *believe*  this try/except is redundant and handled within boto3_conn (inside the module.client call) @jillr or @s-hertel may know the code better...

jillr replied:
This is correct.

tremble commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,284 @@
+  name:
+    description:
+      - Name of the organizational unit
+        This can be full OU path such as "Prod/IT/Service_Desk".
```
```suggestion
        This can be full OU path such as C(Prod/IT/Service_Desk).
```

See https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#linking-within-module-documentation
for more information

tremble commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,284 @@
+
+DOCUMENTATION = """
+module: aws_organization_units
+short_description: Manages AWS Organizations Units
```
```suggestion
short_description: Manages AWS Organizational Units
```

tremble commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,284 @@
+module: aws_organization_units
+short_description: Manages AWS Organizations Units
+description:
+  - Creates and deletes AWS OrganizationalUnits.
```
```suggestion
  - Creates and deletes AWS Organizational Units.
```

tremble reviewed (changes requested):
Mostly down to niggles here.

From experience with things like VPCs I'd still much rather we supported IDs from day one as this enables things like renaming an OU.  Something still *feels* wrong about the use of /Foo/Bar/Baz in the name, but I can understand why you're using it.

jillr commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,284 @@
+  type: str
+  sample: present
+ou:
+  description: The Organization Unit details.
```
```suggestion
  description: The Organizational Unit details.
```

jillr commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,284 @@
+                for ou in page['OrganizationalUnits']:
+                    children_ous.append(dict(
+                        ou,
+                        OrganizationalUnits=self.get_children_ous(ou['Id'], recursive=True),
```
I'm not a big fan of this self recursion. We're a bit limited since boto doesn't give any way to get the entire tree, but it would be nice if this could be broken out differently.

p0tr3c replied:
I've refactored the code, and removed the recursion

jillr reviewed (commented):
Thanks very much for this module @p0tr3c.  As a point of clarification, we normally require integration tests - not unit tests - for AWS modules.  I'd still like to see some integration tests, they would have to be marked as unsupported but we should be able to run them outside of CI as needed.

p0tr3c committed:
Refactor and simplify logic of the code

jillr commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,270 @@
+      - ARN of the organizational unit
+      - ARN can be used to delete OU but not create
+      - Mutually exclusive with c(name)
+    required: true
```
```suggestion
```

p0tr3c replied:
Done

jillr commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,270 @@
+        This can be full OU path such as "Prod/IT/Service_Desk".
+        The lookup always starts from the root node, as the names are not unique within the organization tree.
+      - Mutually exclusive with c(arn)
+    required: true
```
```suggestion
```

p0tr3c replied:
Done

jillr commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,270 @@
+- name: Delete organizational unit
+  aws_organizations:
+    name: Prod
+    state: absent
```
Could you please add an example using `arn` to delete an OU?

p0tr3c replied:
Done

jillr commented on refactor_recursion_notes:
```diff
@@ -0,0 +1,28 @@
+inputs:
```
This file should be removed from the PR

p0tr3c replied:
done

jillr commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,270 @@
+
+EXAMPLES = """
+- name: Create organizational unit
+  aws_organizations:
```
```suggestion
  aws_organization_units:
```

p0tr3c replied:
Done

jillr commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,270 @@
+    state: present
+
+- name: Delete organizational unit
+  aws_organizations:
```
```suggestion
  aws_organization_units:
```

p0tr3c replied:
done

jillr commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,270 @@
+
+    def get_aws_organizational_unit_for_parent(self, ou_name, parent_id):
+        paginator = self.client.get_paginator("list_organizational_units_for_parent")
+        for page in paginator.paginate(ParentId=parent_id):
```
```suggestion
        parent_ous = paginator.paginate(ParentId=parent_id).build_full_result().get('OrganizationalUnits', [])
        for child_ou in parent_ous:
```
You could remove the outer loop here by using the paginator's `.build_full_result()` method and retrieving the OrganizationalUnits from it.

p0tr3c replied:
That looks better thanks

jillr reviewed (changes requested):
This PR also needs an integration test suite added under test/integration/targets/

p0tr3c committed:
Remove notes commited by mistake

p0tr3c committed:
Update examples, correct module names, and use build_full_results

p0tr3c committed:
Fix sanity test errors

p0tr3c committed:
Move arn validation before any call to boto, and require name parameter when state present

p0tr3c commented:
Added basis integration tests to the PR.

p0tr3c committed:
Add basic integration tests

p0tr3c committed:
Add missing aliases file

p0tr3c committed:
Address style comments from PR

p0tr3c committed:
Remove ARN from return doc, this is included under OU dict return value

p0tr3c committed:
Fix issue with missing AWS credentials during client call, and always return the same output on create

p0tr3c committed:
Add more tests, and make them unsupported as they require highest level of privileges

p0tr3c committed:
Remove blank lines

jillr commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,272 @@
+    ou_state = module.params.get('state')
+
+    if ou_arn:
+        if not ou_arn.startswith("arn:aws:organizations::"):
```
There are 3 AWS partitions: aws, aws-cn, and aws-us-gov.  Organizations is not yet available in China but it is in GovCloud, so we should also check for `arn:aws-us-gov:organizations:`.

jillr commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,272 @@
+        try:
+            client = AwsOrganizationalUnit(module, arn=ou_arn)
+        except (BotoCoreError, ClientError) as e:
+            module.fail_json(msg="Boto failure")
```
```suggestion
            module.fail_json_aws(e, msg="Boto failure")
```
Boto/API exceptions should use fail_json_aws

jillr commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,272 @@
+                    if client.delete_aws_organizational_unit():
+                        result['state'] = 'absent'
+                except (BotoCoreError, ClientError, OrganizationalUnitDependencyException) as e:
+                    module.fail_json(msg="Failed to delete organizational unit")
```
```suggestion
                    module.fail_json_aws(e, msg="Failed to delete organizational unit")
```

jillr commented on lib/ansible/modules/cloud/amazon/aws_organization_units.py:
```diff
@@ -0,0 +1,272 @@
+                try:
+                    result['ou'] = client.create_aws_organizational_unit()
+                except (BotoCoreError, ClientError, ParentOrganizationalUnitDoesNotExist) as e:
+                    module.fail_json(msg="Failed to create organizational unit")
```
```suggestion
                    module.fail_json_aws(e, msg="Failed to create organizational unit")
```
and here

jillr reviewed (commented):
Just a couple things left

p0tr3c committed:
Include aws-us-gov as valid arn; Fail via aws specific function

Akasurde commented:
Hi @p0tr3c, Thank you very much for your interest in Ansible. This plugin/module is no longer maintained in this repository and has been migrated to https://github.com/ansible-collections/community.aws<br><br>If you have further questions please stop by IRC or the mailing list:<br><br>* IRC: #ansible on irc.freenode.net<br>* mailing list: https://groups.google.com/forum/#!forum/ansible-project<br><br>needs_info<br>
